src/ui/wxWidgets/PasswordSafeFrame.cpp

```diff
@@ -2432,6 +2432,9 @@ void PasswordSafeFrame::UnlockSafe(bool restoreUI, bool iconizeOnCancel)
       {
         if (ReloadDatabase(scp->GetPassword())) {
           m_sysTray->SetTrayStatus(SystemTray::TrayStatus::UNLOCKED);
+#ifdef __WXMAC__
+          m_InitialTreeDisplayStatusAtOpen = true;
```

Comment:
This change would also be interesting for Linux, so that the tree view is correctly restored after unlocking. Unfortunately, this functionality has been lost with changes to dialog handling.

In the settings (Manage > Options > Display: Initial Tree View) there are three options that affect the tree view once it is displayed.

- Fully collapsed
- Fully expanded
- Same as when last saved

The last option doesn't seem to work correctly anymore. The tree view may be saved incorrectly by calling PasswordSafeFrame::ShowTree and GuiInfo::SaveTreeViewInfo multiple times.

Reply:
I'm not quite sure what you're suggesting here.  Do you want me to remove the ifdef for that line?  On the Mac, I also need the part in SystemTray.cpp to have the view restored at all, otherwise the window comes-up blank.  As for the "last saved" option, it does have an affect for me, but it restores the view to the way it was when the **_file_** was last saved, same as it does for an initial open, since it basically is closing and re-opening the file.  I'm not sure what the original intent was.

Reply:
From my understanding, it should behave as follows regarding the "Initial Tree View" option.
After opening the database, all groups should initially be collapsed, expanded or displayed as last used by the user in the tree view. The "Initial Tree View" option should not have any further effects on the tree view at runtime. Only the changes made by the user to the tree view should be relevant. This means that when switching between views or unlocking the database, the changes should always be saved and restored accordingly.

I was initially under the impression that introducing "m_InitialTreeDisplayStatusAtOpen = true;" fixed the problem of restoring the tree view after an unlock. However, this was due to my set options. It seems like calling "m_guiInfo->RestoreTreeViewInfo(m_tree);" lost with some changes in the past.

The erroneous behavior of the tree view after unlocking the database also results in the loss of focus on the element last selected by the user. (see issue #954)

Reply:
I think what's really happening is _PasswordSafeFrame::SaveAndClearDatabaseOnLock()_ calls _PasswordSafeFrame::ClearAppData()_ which explicitly clears the saved grid and tree info, so there's no state to restore upon unlock.  I'm not sure why that was done. Maybe it was to clear sensitive info from memory?  But, I'm not too sure it accomplishes that.

Reply:
I see the expected behavior of the tree view when calling HideUI(false) in PasswordSafeFrame::IconizeOrHideAndLock() instead of HideUI(true).
However, if there is a breakpoint in GuiInfo::SaveTreeViewInfo, it can be observed that multiple attempts are made to lock the database, which cannot be correct either.
The Show-Hide-Iconize-Lock implementation seems quite opaque to me at the moment.

Reply:
Yes, I understand.  I'm working on a better fix, but I got sidetracked.  I'll update this PR soon.

Reply:
You are right, there are multiple redundant calls to save and restore the tree/grid views.  The last call to save view info happens after the views have been cleared.  GuiInfo::SaveGridViewInfo() has an explicit check at the top (grid->GetNumItems() == 0) that avoids the problem.  As a workaround, I inserted an equivalent check into GuiInfo::SaveTreeViewInfo() and that seems to restore the correct behavior from the user perspective.  I have not tested this on Linux, but I think this might fix #954 as well.  At least, it works as expected on Mac.

I also bumped the version number to 1.17.2 in case new issues appear.

I'm finding the Show-Hide-Iconize-Lock implementation rather confusing as well.  Since any re-work to streamline it would be a complex project, I would like to suggest logging that as a separate issue and letting this PR focus on the user visible behavior.

Reply:
The change to ItemData broke the unit test:

[  FAILED  ] ItemDataTest.PasswordHistory

Reply:
Yes! Commit da6b74881898c91f291eeb75d4ede649e03ab0bd brings back the old behavior to the tree view. :+1:

Reply:
How do I retrieve the log from the test run?
Where is the code for the failed test?

I'm sure it's because I changed the the way the history entry was sorted.  I just discovered that when I transfer my test DB from Mac to Windows, the newly saved history was displayed with oldest at the top.

At this point, I think I should retract this PR, resubmit the changes that seem to be okay, and open issues for the other two.